kinto/plugins/quotas/listener.py

```diff
@@ -0,0 +1,235 @@
+                raise http_error(HTTPInsufficientStorage(),
+                                 errno=ERRORS.FORBIDDEN.value,
+                                 message=message)
+
+        if action == 'create':
+            bucket_info['storage_size'] += new_size
+            if resource_name == 'collection':
+                bucket_info['collection_count'] += 1
+                collection_info['storage_size'] += new_size
+            if resource_name == 'record':
+                bucket_info['record_count'] += 1
+                collection_info['record_count'] += 1
+                collection_info['storage_size'] += new_size
+        elif action == 'update':
+            bucket_info['storage_size'] -= old_size
+            bucket_info['storage_size'] += new_size
+            if resource_name in ('collection', 'record'):
+                collection_info['storage_size'] -= old_size
+                collection_info['storage_size'] += new_size
+        elif action == 'delete':
+            bucket_info['storage_size'] -= old_size
+            if resource_name == 'collection':
+                bucket_info['collection_count'] -= 1
+                # When we delete the collection all the records in it
+                # are deleted without notification.
+                collection_records, _ = storage.get_all('record',
+                                                        collection_uri)
+                for r in collection_records:
+                    old_record_size = record_size(r)
+                    bucket_info['record_count'] -= 1
+                    bucket_info['storage_size'] -= old_record_size
+                    collection_info['record_count'] -= 1
+                    collection_info['storage_size'] -= old_record_size
+                collection_info['storage_size'] -= old_size
+
+            if resource_name == 'record':
+                bucket_info['record_count'] -= 1
+                collection_info['record_count'] -= 1
+                collection_info['storage_size'] -= old_size
+
+    if bucket_max_bytes is not None:
+        if bucket_info['storage_size'] > bucket_max_bytes:
+            message = ("Bucket size exceeded: %d Bytes. "
+                       "MAX_BYTES is %d Bytes" % (
+                           bucket_info['storage_size'], bucket_max_bytes))
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=message)
+
+    if bucket_max_items is not None:
+        if bucket_info['record_count'] > bucket_max_items:
+            message = ("Bucket max items exceeded: %d records. "
+                       "MAX_ITEMS is %d records" % (
+                           bucket_info['record_count'], bucket_max_items))
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=message)
```

Comment:
Actually I find it easier to read like that because it removes a level of indirection and make it explicit that it will raise. I don't think creating a level of indirection with only another function call adds much value here.